aws/resource_aws_route53_health_check.go

```diff
@@ -125,9 +125,10 @@ func resourceAwsRoute53HealthCheck() *schema.Resource {
 				}, true),
 			},
 			"reference_name": {
-				Type:     schema.TypeString,
-				Optional: true,
-				ForceNew: true,
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringLenBetween(0, resource.UniqueIDSuffixLength),
```

Comment:
Hi @timoguin 👋

The maximum length of the [`CallerReference` parameter of the `CreateHealthCheck` API call is 64](https://docs.aws.amazon.com/Route53/latest/APIReference/API_CreateHealthCheck.html#Route53-CreateHealthCheck-request-CallerReference) while the resource creation code sets it to the configuration value hyphen and a 26 character random identifier:

```go
	callerRef := resource.UniqueId()
	if v, ok := d.GetOk("reference_name"); ok {
		callerRef = fmt.Sprintf("%s-%s", v.(string), callerRef)
	}
```

Did you mean the following?

```suggestion
				ValidateFunc: validation.StringLenBetween(0, (64 - resource.UniqueIDSuffixLength - 11), // resource.UniqueIdPrefix and hyphen prefix (-terraform-)
```

Reply:
Yes, I did intend to limit the argument to 27 chars. My mistake.

After perusing the API documentation some more (and https://github.com/terraform-providers/terraform-provider-aws/issues/7248#issuecomment-509663287), the only requirement is that the `CallerRef` is unique for each request.

The one benefit I've found to creating a descriptive `CallerRef` is that the `CreateRoute53HealthCheck` API call can be [logged to CloudTrail](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/logging-using-cloudtrail.html), which contains the `callerReference` in the request parameters.

Example snippet from that doc:
```
            "requestParameters": {
                "callerReference": "2014-05-06 64832",
                "healthCheckConfig": {
                    "iPAddress": "192.0.2.249",
                    "port": 80,
                    "type": "TCP"
                }
            },
```

From the standpoint of security and auditing, it would be beneficial for the `callerReference` to be as descriptive as possible.  This could be improved by allowing as much descriptive text as possible for the `reference_name` argument. Instead of taking up 26 characters for the Terraform-generated suffix, perhaps we could do something like create a trimmed sha1 hash from the current timestamp?

For example, if the generated suffix is `-tf-9e10ab` (first 6 chars from the sha1) then the max length for the  `reference_name` argument could be up to 54 chars.

Full example: `foobar-service-dev-alb-cw-alarm-httpcode-elb-5xx-high-tf-9e10ab` offers descriptive text that could be used to correlate the API request with other other events.

Thoughts?